src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,80 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private static final Map<Class<?>, Class<?>> interfaceImplementations = new HashMap<>();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
```

Comment:
Could you explain to me the necessity of this cache? I don't see such requirement.

Neither do I say it is incorrect. I'm just trying to undestand the reason. We have some injector, that can create class instances and inject their dependencies via Reflection API. But internally it also has some instance cache - which is unknown to the user. So what is the motivation for such implementation detail?

Reply:
I revised the purpose. Originally it should have boosted performance as it returns the object from the map, if it has already been created. But it seems to ruin the Immutability so I got rid of it.